platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -145,51 +156,86 @@ public Hash hash(
+            this.hash = hash;
+            this.nextTask = nextTask;
+            send();
+        }
 
-        private ChunkHashTask out;
+        @Override
+        protected boolean exec() {
+            if (leaf != null) {
+                listener.onLeafHashed(leaf);
+            }
+            listener.onNodeHashed(path, hash);
+            nextTask.send();
+            return true;
+        }
+    }
+
+    class HashHoldingTask extends AbstractTask {
 
         // Input hashes. Some hashes may be null, which indicates they should be loaded from disk
-        private final Hash[] ins;
+        protected final Hash[] ins;
+
+        HashHoldingTask(final ForkJoinPool pool, final int dependencies, final int numHashes) {
+            super(pool, dependencies);
+            ins = numHashes > 0 ? new Hash[numHashes] : null;
+        }
+
+        @Override
+        protected boolean exec() {
+            return true;
+        }
+
+        void setHash(int index, Hash hash) {
+            ins[index] = hash;
+            send();
+        }
+    }
+
+    class ChunkHashTask extends HashHoldingTask {
+
+        private final long path;
+
+        private final int height; // 1 for 3-node chunk, 2 for 7-node chunk, and so on
+
+        private HashHoldingTask out;
 
         // If not null, the task hashes the leaf. If null, the task processes the input hashes
         private VirtualLeafRecord<K, V> leaf;
 
         ChunkHashTask(final ForkJoinPool pool, final long path, final int height) {
-            super(pool, 1 + (1 << height));
+            super(pool, 1 + (1 << height), height > 0 ? 1 << height : 0);
             this.height = height;
             this.path = path;
-            this.ins = new Hash[1 << height];
         }
 
-        void setOut(final ChunkHashTask out) {
+        void setOut(final HashHoldingTask out) {
             this.out = out;
-            assert path == 0 || Path.getRank(path) - out.height == Path.getRank(out.path)
-                    : "setOut " + path + " " + height + " " + out.path;
             send();
         }
 
-        void setData(final VirtualLeafRecord<K, V> leaf) {
-            assert leaf == null || path == leaf.getPath();
-            assert leaf == null || height == 1;
-            assert leaf != null || out != null;
-            if (leaf == null) {
-                out.setHash(getIndexInOut(), null);
-            } else {
-                this.leaf = leaf;
-                send(); // left hash dependency
-                send(); // right hash dependency
-            }
+        void setLeaf(final VirtualLeafRecord<K, V> leaf) {
+            assert leaf != null && path == leaf.getPath() && height == 0;
+            this.leaf = leaf;
+            send();
         }
 
-        void setHash(final int index, final Hash hash) {
-            assert index >= 0 && index < (1 << height);
-            ins[index] = hash;
-            send();
+        void complete() {
+            out.send();
```

Comment:
It would be great to have a couple assertions here. For "leaf" tasks make sure no leaf data is set. For "internal" tasks make sure no input hashes are set.